src/demo/java/org/junitpioneer/jupiter/EnvironmentVariablesExtensionDemo.java

```diff
@@ -60,4 +65,62 @@ void clearedAtClasslevel() {
 	}
 	// end::environment_using_at_class_level[]
 
+	// tag::environment_method_restore_test[]
+	@ParameterizedTest
+	@ValueSource(strings = { "foo", "bar" })
+	@RestoreEnvironmentVariables
+	void parameterizedTest(String value) {
+		setEnvVar("some parameterized property", value);
+		setEnvVar("some other dynamic property", "my code calculates somehow");
+	}
+	// end::environment_method_restore_test[]
+
+	// tag::environment_class_restore[]
+	@RestoreSystemProperties
+	class MyEnvironmentVariableRestoreTest {
+
+		@BeforeAll
+		public void beforeAll() {
+			setEnvVar("A", "A value");
+		}
+
+		@BeforeEach
+		public void beforeEach() {
+			setEnvVar("B", "B value");
+		}
+
+		@Test
+		void isolatedTest1() {
+			setEnvVar("C", "C value");
+		}
+
+		@Test
+		void isolatedTest2() {
+			// A & B are visible, C is not
+		}
+
+	}
+
+	class SomeOtherTest {
+		// Changes to A, B & C have been restored to their values prior to the above test
```

Comment:
Not sure if I can follow, why do you have to adapt the `TestInstance.Lifecycle`? Envvars A and B are set before all/each, only C comes from within a test?

Reply:
The issue is that `@BeforeAll` cannot be used on a non-static method unless the test is marked as `Lifecycle.PER_CLASS`, and a static method cannot be used in a non-static inner class on JDK 11.  I'd rather not reveal the `Lifecycle.PER_CLASS` annotation in the docs, since normally it is not required.

I could work around this by either  making two separate code snippets, or using separate files.  I think I'm going to make two code snippets and bring them into the docs separately.

Reply:
I also tried applying `Lifecycle.PER_CLASS` on a wrapping class and it is not inherited in that way (apparently only via class inheritance).

I did come up w/ a solution here that splits the demo into two pieces so I can clip out the annotations that are not relevant to the demo.